[JSC] Use simdutf for fromBase64

Constellation opened the pull request:
#### 33c82ad83b428e694d506ea257ed4e5277da9dfc
<pre>
[JSC] Use simdutf for fromBase64
<a href="https://bugs.webkit.org/show_bug.cgi?id=277584">https://bugs.webkit.org/show_bug.cgi?id=277584</a>
<a href="https://rdar.apple.com/133101730">rdar://133101730</a>

Reviewed by Yijia Huang.

This patch integrates simdutf into fromBase64.

1. When &quot;strict&quot; or &quot;stop-before-partial&quot; options are specified, we use fallback implementation[1]. Once they are implemented, we will switch to that.
2. Upgrade simdutf to fix issues found during implementing TC39 Base64 features.

[1]: <a href="https://github.com/simdutf/simdutf/issues/440">https://github.com/simdutf/simdutf/issues/440</a>

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.cpp: Removed.
* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/simdutf/simdutf_impl.cpp:
* Source/WTF/wtf/text/Base64.cpp:
(WTF::fromBase64SlowImpl):
(WTF::fromBase64Impl):
(WTF::fromBase64):
(WTF::maxLengthFromBase64):
* Source/WTF/wtf/text/Base64.h:

Canonical link: <a href="https://commits.webkit.org/281899@main">https://commits.webkit.org/281899@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c52661c214e3b61bf5c466c2b9c41f0d9ff3b734

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61390 "403 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40750 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13974 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65342 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11938 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48428 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12213 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/49590 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8288 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64330 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37881 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53174 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30433 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34542 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10402 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10851 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/54493 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56350 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10702 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67073 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60638 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5336 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/10476 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/56966 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5359 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53139 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57181 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4406 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/82406 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9232 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36554 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14380 "Found 1466 new JSC stress test failures: microbenchmarks/memcpy-wasm-large.js.no-cjit-validate-phases, microbenchmarks/memcpy-wasm-medium.js.no-cjit-validate-phases, microbenchmarks/memcpy-wasm-small.js.no-cjit-validate-phases, microbenchmarks/memcpy-wasm.js.no-cjit-validate-phases, microbenchmarks/wasm-cc-int-to-int.js.default-wasm, stress/proxy-get-and-set-recursion-stack-overflow.js.eager-jettison-no-cjit, stress/proxy-get-and-set-recursion-stack-overflow.js.no-cjit-collect-continuously, stress/proxy-get-and-set-recursion-stack-overflow.js.no-cjit-validate-phases, stress/proxy-set.js.bytecode-cache, stress/proxy-set.js.default ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37637 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38731 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37381 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.cpp:
```diff
@@ -87,18 +87,18 @@ JSC_DEFINE_HOST_FUNCTION(uint8ArrayConstructorFromBase64, (JSGlobalObject* globa
 
-    auto [shouldThrowError, readLength, writeData] = fromBase64(view, std::numeric_limits<size_t>::max(), alphabet, lastChunkHandling);
-    if (shouldThrowError == FromBase64ShouldThrowError::Yes)
+    Vector<uint8_t, 128> output;
```
Why don't we just do `Vector<uint8_t, maxLengthFromBase64(view)> output;` here?

Constellation replied:
`maxLengthFromBase64` depends on length, so it is not constant size.

hyjorc1 replied:
Oops, right.

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Use simdutf for fromBase64
https://bugs.webkit.org/show_bug.cgi?id=277584
rdar://133101730

Reviewed by Yijia Huang.

This patch integrates simdutf into fromBase64.

1. When "strict" or "stop-before-partial" options are specified, we use fallback implementation[1]. Once they are implemented, we will switch to that.
2. Upgrade simdutf to fix issues found during implementing TC39 Base64 features.

[1]: https://github.com/simdutf/simdutf/issues/440

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.cpp: Removed.
* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/simdutf/simdutf_impl.cpp:
* Source/WTF/wtf/text/Base64.cpp:
(WTF::fromBase64SlowImpl):
(WTF::fromBase64Impl):
(WTF::fromBase64):
(WTF::maxLengthFromBase64):
* Source/WTF/wtf/text/Base64.h:

Canonical link: https://commits.webkit.org/281899@main
